wait.go

```diff
@@ -246,36 +255,78 @@ type httpWaiter struct {
 
 // Wait waits for something to be accepting HTTP requests.
 func (w httpWaiter) Wait(ctx context.Context) error {
-	for {
-		if err := checkContextDone(ctx, w.logger, w.url); err != nil {
-			return err
-		}
+	if err := waitAndRetry(ctx, w.waitOnce); err != nil {
+		return serrors.Errorf("connecting to %q: %w", w.url, err)
+	}
 
-		res, _ := w.client.Get(w.url)
-		if res == nil {
-			continue
-		}
-		res.Body.Close()
+	return nil
+}
 
-		if res.StatusCode < http.StatusInternalServerError {
-			w.logger.DebugContext(ctx, "established connection to address",
-				"address", w.url,
-			)
-			return nil
-		}
+func (w httpWaiter) waitOnce(ctx context.Context) waitResult {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, w.url, nil)
+	if err != nil {
+		return waitResult{FatalErr: serrors.WithStack(err)}
 	}
-}
 
-// checkContextDone checks if the provided context is done, and returns
-// an error if it is.
-func checkContextDone(ctx context.Context, logger *slog.Logger, addr string) error {
-	select {
-	case <-ctx.Done():
-		logger.DebugContext(ctx, "failed to establish connection to address",
-			"address", addr,
+	res, err := w.client.Do(req)
+	if err != nil {
+		w.logger.DebugContext(ctx, "failed to connect to address",
+			"address", w.url,
+			"error", err,
 		)
-		return serrors.Errorf("timed out connecting to %q", addr)
-	default:
-		return nil
+		return waitResult{Ok: false}
+	}
+	res.Body.Close()
+
+	w.logger.DebugContext(ctx, "received status from address",
+		"address", w.url,
+		"http_status", res.StatusCode,
+	)
+	return waitResult{
+		Ok: res.StatusCode < http.StatusInternalServerError,
+	}
+}
+
+type waitResult struct {
+	// Ok returns whether the operation was successful.
+	Ok bool
+	// FatalErr indicates the retry loops should end immediately.
+	FatalErr error
+}
+
+func waitAndRetry(
+	ctx context.Context,
+	waitFunc func(context.Context) waitResult,
+) error {
+	// In an ideal world, this is configurable. Without making any breaking
+	// changes, this is probably a reasonable interval compared to the previous
+	// behavior of "make a new request as soon as the old one finished".
+	interval := 100 * time.Millisecond
```

Comment:
One way to pass this down would be attaching a value to the context, which the user can specify with a vardiac option function. 

Not really necessary right now, but just a thought.

Reply:
There are a couple ways to do this that are "technically" a breaking change but not "practically" a breaking change that I think we could pursue. That said, I'm cool punting on it until we prove we need it either way.